### PR TITLE
DATAUP-463: job manager job ID deduplication

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

Added a new function to deduplication job IDs in a list, removing any None or '' values first.

Also made some generic fixes, like swapping out repeated strings for constants and updating method docs

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-463
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and seeing that there are no noticeable changes

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook

### CODE REVIEW

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 1]
```diff
@@ -25,7 +23,13 @@
 __version__ = "0.0.1"
 
 TERMINAL_STATES = ["completed", "terminated", "error"]
-EXCLUDED_JOB_STATE_FIELDS = ["authstrat", "job_input", "condor_job_ads"]
+EXCLUDED_JOB_STATE_FIELDS = [
+    "authstrat",
+    "condor_job_ads",
+    "job_input",
+    "scheduler_type",
+    "scheduler_id",
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:31]: add in two extra unneeded job fields

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 2]
```diff
@@ -88,12 +93,17 @@ def initialize_jobs(self):
...
             )
+            refresh_state = 1 if status not in TERMINAL_STATES else 0
             self._running_jobs[job_id] = {
-                "refresh": 1
-                if status not in ["completed", "errored", "terminated"]
-                else 0,
+                "refresh": refresh_state,
                 "job": job,
             }
+            # if the job is in a terminal state, add it to the _completed_job_states index
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:101]: initialisation previously did not include adding jobs to the `_completed_job_states` index, so one had to wait until there was a status update to populate the index properly. This fixes that problem.

File: src/biokbase/narrative/jobs/jobmanager.py [hunk 3]
```diff
@@ -130,6 +140,34 @@ def _create_jobs(self, job_ids):
                 # be set to True.
                 self._running_jobs[job_id] = {"refresh": 0, "job": job}
 
+    def _check_job_list(self, job_id_list: List[str] = []):
```
Comment (ialarmedalien) [inline src/biokbase/narrative/jobs/jobmanager.py:143]: New function for job ID cleaning and deduping.

File: src/biokbase/narrative/tests/narrative_mock/mockclients.py [hunk 4]
```diff
@@ -1,6 +1,9 @@
 from ..util import ConfigTests
 from biokbase.workspace.baseclient import ServerError
 
+RANDOM_DATE = "2018-08-10T16:47:36+0000"
+RANDOM_TYPE = "ModuleA.TypeA-1.0"
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/narrative_mock/mockclients.py:5]: string => constant substitution

File: src/biokbase/narrative/tests/test_jobcomm.py [hunk 5]
```diff
@@ -520,7 +526,7 @@ def test_request_no_req(self):
         rq_msg2 = {"msg_id": "some_other_id", "content": {"data": {}}}
         for msg in [rq_msg, rq_msg2]:
             with self.assertRaises(ValueError) as e:
-                JobRequest(rq_msg)
+                JobRequest(msg)
```
Comment (ialarmedalien) [inline src/biokbase/narrative/tests/test_jobcomm.py:529]: fix this test so it doesn't do the same thing twice

File: src/biokbase/narrative/jobs/jobmanager.py (hunk 3 above)
Comment (briehl) [reply src/biokbase/narrative/jobs/jobmanager.py:135]: How often do we see duplication? Or is this just being cautious?
I guess I could see the lookup for an individual job, plus that job again as the child of some other job.

**Review by briehl — APPROVED** [review]
Looks fine, just a question in the jobmanager.

File: src/biokbase/narrative/jobs/jobmanager.py (hunk 3 above)
Comment (ialarmedalien) [reply src/biokbase/narrative/jobs/jobmanager.py:135]: mostly just being cautious -- the main example I thought of was the child job case you mention.

**Pull request merged**